Parsing an invalid FIX message could cause infinite loop and OutOfMemory

**Out of memory**
Client send us enormous  amount of hearbeats.
org.apache.mina.filter.codec.ProtocolCodecFilter$ProtocolDecoderOutputImpl messageQueue in the heapdump contains  9854469 unprocessed heartbeats
Also SimpleBufferAllocator$SimpleBuffer contains unprocessed heartbeats
`8=FIX.4.4.9=082.35=0.49=xxxxxxxxxxxx.xxxxxxxx.56=cServer.57=QUOTE.52=20211021-21:44:48.318.34=38046.10=060..8=FIX.4.4.9=082.35=0.49=xxxxxxxxxxxx.xxxxxxxx.56=cServer.57=QUOTE.52=20211021-21:44:48.318.34=38047.10=061..8=FIX.4.4.9=082.35=0.49=xxxxxxxxxxxx.xxxxxxxx.56=cServer.57=QUOTE.52=20211021-21:44:48.318.34=38048.10=062..8=FIX.4.4.9=082.35=0.49=xxxxxxxxxxxx.xxxxxxxx.56=cServer.57=QUOTE.52=20211021-21:44:48.319.34=38049.10=064..8=FIX.4.4.9=082.35=0.49=xxxxxxxxxxxx.xxxxxxxx.56=cServer.57=QUOTE.52=20211021-21:44:48.319.34=38050.10=056..8=FIX.4.4.9=082.35=0.49=xxxxxxxxxxxx.xxxxxxxx.56=cServer.57=QUOTE.52=20211021-21:44:48.319.34=38051.10=057..8=FIX.4.4.9=082.35=0.49=xxxxxxxxxxxx.xxxxxxxx.56=cServer.57=QUOTE.52=20211021-21:44:48.319.34=38052.10=058..8=FIX.4.4.9=082.35=0.49=xxxxxxxxxxxx.xxxxxxxx.56=cServer.57=QUOTE.52=20211021-21:44:48.319.34=38053.10=059..8=FIX.4.4.9=082.35=0.49=xxxxxxxxxxxx.xxxxxxxx.56=cServer.57=QUOTE.52=20211021-21:44:48.319.34=38054.10=060..8=FIX.4.4.9=082.35=0.49=xxxxxxxxxxxx.xxxxxxxx.56=cServer.57=QUOTE.52=202`

Verification of sequence number is disabled, because we had OOM in case when client send us wrong sequence number.
The messageQueue contains repeated heartbeats. It looks like client send 20 generated heartbeats many times and kill our application in 1-2 minutes.

Some packages had mistake like `Length format error in message (last character: =): 8=FIX.4.4.9=8=FIX.4.4.9=08`

**Step to reproduce**

`        SessionID currentSessionId = (SessionID) ReflectionUtils.getField(field, connectionInitiator);
        
        final Session session = Session.lookupSession(currentSessionId);
        
        MemoryPoolMXBean mxBean = ManagementFactory.getMemoryPoolMXBeans().stream()
                .filter((MemoryPoolMXBean memoryPoolMXBean) -> memoryPoolMXBean.getName().equals("G1 Old Gen"))
                .findFirst().orElseThrow();
        long counter = 0;
        Heartbeat heartbeat = new Heartbeat();
        List<String> out = new ArrayList<>();
        for(int i = 0; i < 20; i++) {
            heartbeat.getHeader().removeField(PossDupFlag.FIELD);
            heartbeat.getHeader().removeField(OrigSendingTime.FIELD);
            initializeHeader(heartbeat.getHeader(), currentSessionId, i + 2);
            out.add(heartbeat.toString());
        }

        StringBuilder builder = new StringBuilder();
        String s = heartbeat.toString().substring(0, 13);
        for(int i = 0; i < 100; i++) {
            builder.append(s);
        }
        out.add(s.toString());
        
        Responder responder = session.getResponder();
        
        while (true) {
            if (mxBean.getUsage().getUsed() < 128L * 1024 * 1024) {
                if (session.isLoggedOn()) {

                    for (String v : out) {
                        responder.send(v);
                    }
                } else {
                    break;
                }
            } else {
                System.gc();
            }
        }
        LOGGER.info("Counter {}", counter);
    }
    
    private static void initializeHeader(Message.Header header, SessionID sessionID, int seq) {
        header.setString(BeginString.FIELD, sessionID.getBeginString());
        header.setString(SenderCompID.FIELD, sessionID.getSenderCompID());
        optionallySetID(header, SenderSubID.FIELD, sessionID.getSenderSubID());
        optionallySetID(header, SenderLocationID.FIELD, sessionID.getSenderLocationID());
        header.setString(TargetCompID.FIELD, sessionID.getTargetCompID());
        optionallySetID(header, TargetSubID.FIELD, sessionID.getTargetSubID());
        optionallySetID(header, TargetLocationID.FIELD, sessionID.getTargetLocationID());
        header.setInt(MsgSeqNum.FIELD, seq);
        insertSendingTime(header);
    }

    private static void optionallySetID(Header header, int field, String value) {
        if (!value.equals(SessionID.NOT_SET)) {
            header.setString(field, value);
        }
    }

    private static void insertSendingTime(Message.Header header) {
        header.setUtcTimeStamp(SendingTime.FIELD, SystemTime.getLocalDateTime(), UtcTimestampPrecision.SECONDS);
    }
}`

**Expected behavior**
Is it possible to add some rate limiter for income packages?
Add limit on broken messages on transport layer?
I tried `        SocketAcceptor socketAcceptor = SocketAcceptor.newBuilder()
                .withApplication(application)
                .withLogFactory(logFactory)
                .withMessageFactory(getMessageFactory())
                .withMessageStoreFactory(createMessageStoreFactory(settings))
                .withQueueWatermarks(10, 10000)
                .withSettings(settings)
                .build();`
but it wouldn't help.
**system information:**
 - OS: Linux 3.10.0-1160.36.2.el7.x86_64
 - Java version OpenJDK 64-Bit Server VM  11
 - QFJ Version 2.1.1

My fix only was

```
handleError(in, position, "Length format error in message (last character: " + (char)ch + "): " + messageString,
                                 false);
```
So I only put `position` instead of `in.position() + 1`. Do you think the other changes are needed? Why should we skip one more character?

My unit test works with your message as well but your unit test does not seem to work with your proposed fix?!

I got: `java.lang.AssertionError: wrong decoder result expected:<OK> but was:<NEED_DATA>`
Or were you just trying to show that it does not get into an infinite loop?

> My fix only was
> 
> ```
> handleError(in, position, "Length format error in message (last character: " + (char)ch + "): " + messageString,
>                                  false);
> ```
> 
> So I only put `position` instead of `in.position() + 1`. Do you think the other changes are needed? Why should we skip one more character?
> 
> My unit test works with your message as well but your unit test does not seem to work with your proposed fix?!
> 
> I got: `java.lang.AssertionError: wrong decoder result expected:<OK> but was:<NEED_DATA>` Or were you just trying to show that it does not get into an infinite loop?
Try this https://github.com/quickfix-j/quickfixj/files/7511786/FIXMessageDecoder.patch.txt

--- a/quickfixj-core/src/main/java/quickfix/mina/message/FIXMessageDecoder.java
+++ b/quickfixj-core/src/main/java/quickfix/mina/message/FIXMessageDecoder.java
@@ -170,8 +170,8 @@ public class FIXMessageDecoder implements MessageDecoder {
                     } else {
                         if (position < in.limit()) { // if data remains
                             String messageString = getMessageStringForError(in);
-                            handleError(in, in.position() + 1, "Length format error in message (last character: " + (char)ch + "): " + messageString,
-                                false);
+                            handleError(in, position, "Length format error in message (last character: " + (char) ch + "): " + messageString,
+                                    false);
                             continue;
                         } else {
                             break;

--- a/quickfixj-core/src/test/java/quickfix/mina/message/FIXMessageDecoderTest.java
+++ b/quickfixj-core/src/test/java/quickfix/mina/message/FIXMessageDecoderTest.java
@@ -490,4 +490,30 @@ public class FIXMessageDecoderTest {
         setUpBuffer(message);
         assertMessageFound(goodMessage);
     }
+    
+    /**
+     * Several bad messages after each other should not send the decoder in an
+     * infinite loop. https://github.com/quickfix-j/quickfixj/issues/432
+     */
+    @Test(timeout = 1000)
+    public void testLengthFormatError() throws Exception {
+        String badMessages = "8=FIX.4.4\0019=058=\0018=FIX.4.4\0019=058=\0018=FIX.4.4\0019=058=\0018=FIX.4.4\0019=058=\001";
+        String goodMessage = "8=FIX.4.4\0019=12\00135=Y\001108=30\00110=037\001";
+        setUpBuffer(badMessages + goodMessage + badMessages + goodMessage);
+        assertMessageFound(goodMessage, 2);
+    }
+
+    /**
+     * Several bad messages after each other should not send the decoder in an
+     * infinite loop. https://github.com/quickfix-j/quickfixj/issues/432
+     */
+    @Test(timeout = 1000)
+    public void testLengthFormatError2() throws Exception {
+        decoder = new FIXMessageDecoder("UTF-16");
+        setUpBuffer("8=FIX.4.2\0019=128=FIX.4.2\0019=8=FIX.4.2\0019=128="
+                + "FIX.4.2\0019=8=FIX.4.2\0019=12\00135=X\001108=30\00110=049\001");
+        MessageDecoderResult decoderResult = decoder.decode(null, buffer, decoderOutput);
+        assertEquals("wrong decoder result", MessageDecoderResult.OK, decoderResult);
+        assertEquals("Wrong encoding", 14397, (int) decoderOutput.getMessage().charAt(0));
+    }
 }

